src/global_sequences_viewer.cpp

```diff
@@ -0,0 +1,169 @@
+
+        status = registerManager(managerId, stetefulModelInstance->getSequenceManager().get());
+        if (status.getCode() != ovms::StatusCode::OK)
+            return status;
+    }
+    updateThreadInterval();
+    return ovms::StatusCode::OK;
+}
+
+void GlobalSequencesViewer::updateThreadInterval() {
+    uint32_t lowestHalfTimeoutInterval = DEFAULT_SEQUENCE_TIMEOUT_SECONDS / 2;
+    for (auto const& [key, val] : registeredSequenceManagers) {
+        auto sequenceManager = val;
+        uint32_t newInterval = sequenceManager->getTimeout() / 2;
+        if (newInterval > 0 && newInterval < lowestHalfTimeoutInterval)
+            lowestHalfTimeoutInterval = newInterval;
+    }
+
+    sequenceWatcherIntervalSec = lowestHalfTimeoutInterval;
+    SPDLOG_LOGGER_DEBUG(modelmanager_logger, "Sequence watcher thread interval set to {}", sequenceWatcherIntervalSec);
+}
+
+ovms::Status GlobalSequencesViewer::registerManager(std::string managerId, SequenceManager* sequenceManager) {
+    std::unique_lock<std::mutex> viewerLock(mutex);
+    if (registeredSequenceManagers.count(managerId)) {
+        SPDLOG_LOGGER_DEBUG(sequence_manager_logger, "Sequence manager {} already exists", managerId);
+        return StatusCode::SEQUENCE_ALREADY_EXISTS;
+    } else {
+        SPDLOG_LOGGER_DEBUG(sequence_manager_logger, "Sequence manager {} added", managerId);
+        registeredSequenceManagers.emplace(managerId, sequenceManager);
+    }
+
+    return StatusCode::OK;
+}
+
+ovms::Status GlobalSequencesViewer::unregisterManager(std::string managerId) {
+    std::unique_lock<std::mutex> viewerLock(mutex);
+    if (registeredSequenceManagers.count(managerId)) {
+        SPDLOG_LOGGER_DEBUG(sequence_manager_logger, "Sequence manager {} removed", managerId);
+        registeredSequenceManagers.erase(managerId);
+    } else {
+        SPDLOG_LOGGER_DEBUG(sequence_manager_logger, "Sequence manager {} already exists", managerId);
+        return StatusCode::SEQUENCE_MISSING;
+    }
+    return StatusCode::OK;
+}
+
+ovms::Status GlobalSequencesViewer::removeTimedOutSequences() {
+    std::unique_lock<std::mutex> viewerLock(mutex);
+    for (auto it = registeredSequenceManagers.begin(); it != registeredSequenceManagers.end();) {
+        auto sequenceManager = it->second;
+        auto status = sequenceManager->removeTimedOutSequences();
+        it++;
+        if (status.getCode() != ovms::StatusCode::OK)
+            return status;
+    }
+
+    return ovms::StatusCode::OK;
+}
+
+void GlobalSequencesViewer::sequenceWatcher(std::future<void> exit) {
+    SPDLOG_LOGGER_INFO(modelmanager_logger, "Started sequence timeout watcher thread");
+
+    while (exit.wait_for(std::chrono::milliseconds(1)) == std::future_status::timeout) {
+        std::this_thread::sleep_for(std::chrono::seconds(sequenceWatcherIntervalSec));
+        SPDLOG_LOGGER_DEBUG(modelmanager_logger, "Sequence watcher thread check cycle begin");
+
+        removeTimedOutSequences();
```

Comment:
This call is locking the viewer, but do adding, removing and reloading models?

Reply:
register and unregister does lock it.